botocove/cove_host_account.py

```diff
@@ -81,175 +94,229 @@ def get_cove_sessions(self) -> List[CoveSessionInformation]:
-            return assuming_session.client(
-                service_name=clientname,
-                config=Config(max_pool_connections=self.thread_workers),
-            )
-        logger.info("No Boto3 session argument: using credential chain")
-        return boto3.client(
-            service_name=clientname,
-            config=Config(max_pool_connections=self.thread_workers),
-        )
+    # Is target_accounts part of the public interface? I see it used in tests but
+    # I don't see why a user would refer to it.
+    @property
+    def target_accounts(self) -> Set[str]:
+        return set(self.target_account_map)
 
-    def _get_boto3_org_client(
-        self, assuming_session: Optional[Session]
-    ) -> OrganizationsClient:
-        client: OrganizationsClient = self._get_boto3_client(
-            "organizations", assuming_session
-        )
-        return client
-
-    def _get_boto3_sts_client(self, assuming_session: Optional[Session]) -> STSClient:
-        client: STSClient = self._get_boto3_client("sts", assuming_session)
-        return client
-
-    def _resolve_target_accounts(self, target_ids: Optional[List[str]]) -> Set[str]:
-        accounts_to_ignore = self._gather_ignored_accounts()
-        logger.info(f"Ignoring account IDs: {accounts_to_ignore=}")
-        accounts_to_target = self._gather_target_accounts(target_ids)
-        final_accounts: Set = accounts_to_target - accounts_to_ignore
-        if len(final_accounts) < 1:
-            raise ValueError(
-                "There are no eligible account ids to run decorated func against"
-            )
-        return final_accounts
-
-    def _gather_ignored_accounts(self) -> Set[str]:
-        ignored_accounts = {self.host_account_id}
-
-        if self.provided_ignore_ids:
-            accs, ous = self._get_validated_ids(self.provided_ignore_ids)
-            ignored_accounts.update(accs)
-            if ous:
-                accs_from_ous = self._get_all_accounts_by_organization_units(ous)
-                ignored_accounts.update(accs_from_ous)
-
-        return ignored_accounts
-
-    def _gather_target_accounts(self, targets: Optional[List[str]]) -> Set[str]:
-        if targets:
-            accs, ous = self._get_validated_ids(targets)
-            if ous:
-                accs_from_ous = self._get_all_accounts_by_organization_units(ous)
-                accs.extend(accs_from_ous)
-            return set(accs)
-        else:
-            # No target_ids passed, getting all accounts in org
-            return self._get_active_org_accounts()
 
-    def _get_validated_ids(self, ids: List[str]) -> Tuple[List[str], List[str]]:
+def _host_account_is_in_organization(client: OrganizationsClient) -> bool:
+    try:
+        client.describe_organization()
+        return True
+    except ClientError as error:
+        if _organizations_service_not_in_use(error):
+            return False
+        raise error
```

Comment:
Blocking problem here:

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/organizations.html#Organizations.Client.describe_organization

> This operation can be called from any account in the organization.

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/organizations.html#Organizations.Client.list_accounts


> This operation can be called only from the organization's management account or by a member account that is a delegated administrator for an Amazon Web Services service. 

being in an org != allowed to call list_accounts - this breaks the behaviour of using a non-org-master account within an org


Nitpick:
```suggestion
def _host_account_is_in_organization(client: OrganizationsClient) -> bool:
    try:
        client.describe_organization()
        return True
    except ClientError as error:
        if error.response["Error"]["Code"] == "AWSOrganizationsNotInUseException"
            return False
        raise error
```

think the extra func call here is a bit misleading as to something more going on

Reply:
> being in an org != allowed to call list_accounts - this breaks the behaviour of using a non-org-master account within an org

Yes, good catch. But I'm confused now. If you are running cove in an org, why would you run it anywhere other than the master account?

I'm having trouble understanding the use case. I can see value in setting up a "jump" account separate from the master account. Is that what you had in mind?

Reply:
> Nitpick:

Agreed. It is clearer to inline the error code check in this case.

Organizations has a specific error code for the condition we care about so encapsulating the check in a function is less readable.

I had gotten into the habit of using boolean helper functions to check Boto3 client errors because some services return quite generic errors. [CloudFormation's ValidationError](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_DescribeStacks.html) for example) covers many conditions and to find out exactly what went wrong you need to parse the message.

Reply:
> Yes, good catch. But I'm confused now. If you are running cove in an org, why would you run it anywhere other than the master account?
> 
> I'm having trouble understanding the use case. I can see value in setting up a "jump" account separate from the master account. Is that what you had in mind?

This is something suggested as best practice: https://docs.aws.amazon.com/organizations/latest/userguide/orgs_best-practices_mgmt-acct.html#best-practices_mgmt-use - I did think about outlining in the readme which I should probably do!

A suggested solution for using Cove in automation would be to deploy a read-only role [or appopriate policy] trusting a security audit account or similar across the org.